amaranth/vendor/gowin.py

```diff
@@ -0,0 +1,548 @@
+from abc import abstractproperty
+
+from ..hdl import *
+from ..lib.cdc import ResetSynchronizer
+from ..build import *
+
+# Acknowledgments:
+#   Parts of this file originate from https://github.com/tcjie/Gowin
+
+
+__all__ = ["GowinPlatform"]
+
+
+class GowinPlatform(TemplatedPlatform):
+    """
+    .. rubric:: Apicula toolchain
+
+    Required tools:
+        * ``yosys``
+        * ``nextpnr-gowin``
+        * ``gowin_pack``
+
+    The environment is populated by running the script specified in the environment variable
+    ``AMARANTH_ENV_APICULA``, if present.
+
+    Build products:
+        * ``{{name}}.fs``: binary bitstream.
+
+    .. rubric:: Gowin toolchain
+
+    Required tools:
+        * ``gw_sh``
+
+    The environment is populated by running the script specified in the environment variable
+    ``AMARANTH_ENV_GOWIN``, if present.
+
+    Build products:
+        * ``{{name}}.fs``: binary bitstream.
+
+    """
+
+    toolchain = None # selected when creating platform
+
+    device    = abstractproperty()
+    series    = abstractproperty()
+    subseries = abstractproperty()
+    voltage   = abstractproperty()
+    size      = abstractproperty()
+    package   = abstractproperty()
+    speed     = abstractproperty()
+    osc_div   = abstractproperty()
```

Comment:
Done.

Reply:
Oh, should it be called `osc_frequency` instead of `osc_freq`? And is there a convention on the unit of `osc_freq*` ? It is convenient to use MHz as unit and allow floats. One might also argue that only integers should be allowed and use Hz as base unit.

Reply:
It should be called `osc_frequency` and the unit must be Hz. Take a look at `LatticeMachXO2Or3LPlatform`, which can be used as a template here.

Reply:
Also, I think a comment on this line might've went missing. It shouldn't be an `abstractproperty` because this means a design that doesn't use the built-in oscillator at all cannot be instantiated (I'm actually hitting this issue). Rather, you can remove the property, and have `AttributeError` propagate when a clock constraint is requested on a design without `osc_frequency` set.